src/pybind/mgr/diskprediction/agent/predict/prediction.py

```diff
@@ -0,0 +1,308 @@
+from __future__ import absolute_import
+
+import datetime
+import json
+import os
+import time
+
+from .. import BaseAgent
+from ...common.db import DB_API
+
+
+PREDICTION_FILE = '/var/tmp/disk_prediction.json'
+
+test_json = '{"results":[{"statement_id":0,"series":[{"name":"sai_disk_prediction","columns":["time","cluster_domain_id","confidence","disk_domain_id","disk_model","disk_name","disk_serial_number","disk_type","disk_vendor","host_domain_id","life_expectancy","life_expectancy_day","near_failure","predicted","primary_key"],"values":[["2018-05-25T01:26:18.231490725Z","dpCluster",100,"55cd2e404b7ee6d3","INTEL SSDSC2BP480G4","MegaraidDisk-0","BTJR516601GW480BGN","5","","da24c5fac654244dccffeb6b564b139b",24,730,"Good",1527211578228188894,"dpCluster-da24c5fac654244dccffeb6b564b139b-55cd2e404b7ee6d3"]]}]}]}'
+TIME_DAYS = 24*60*60
+TIME_WEEK = TIME_DAYS * 7
+
+
+class Prediction_Agent(BaseAgent):
+
+    measurement = 'sai_disk_prediction'
+
+    @staticmethod
+    def _get_disk_type(is_ssd, vendor, model):
+        """ return type:
+            0: "Unknown", 1: "HDD",
+            2: "SSD",     3: "SSD NVME",
+            4: "SSD SAS", 5: "SSD SATA",
+            6: "HDD SAS", 7: "HDD SATA"
+        """
+        if is_ssd:
+            if vendor:
+                disk_type = 4
+            elif model:
+                disk_type = 5
+            else:
+                disk_type = 2
+        else:
+            if vendor:
+                disk_type = 6
+            elif model:
+                disk_type = 7
+            else:
+                disk_type = 1
+        return disk_type
+
+    def _store_prediction_result(self, result):
+        self._ceph_context._prediction_result = result
+
+    def _parse_prediction_data(self, host_domain_id, disk_domain_id):
+        result = {}
+
+        sql = "SELECT * FROM \"%s\" WHERE (%s) ORDER BY time DESC LIMIT 1"
+        where = "host_domain_id=\'%s\' AND disk_domain_id=\'%s\'" % (host_domain_id, disk_domain_id)

```

Comment:
Is this passing an arbitrary SQL query to the server?  Is there something preventing this from returning data for other devices from other users?  (!!!)

Reply:
@liewegas Data of different users are keep in different namespace at cloud. This is controlled by the diskprediction server config setting.